### PR TITLE
A tale of servant clients

### PR DESCRIPTION
**Pull request by divarvel**
- [x] links
- [x] `debugLayout` output
- [x] add imports (or at least explain where does `(>>>)` come from
- [x] refactor a bit the `Application` monad
- [x] test the code examples

### CODE REVIEW

File: posts/2018-12-27-a-tale-of-servant-clients.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+In the codebase I am working on, I went with a `>>>` / `withParam` combo,
+and I am quite happy with the result. Maybe the same could be achieved with
+lenses, but I have not tried it yet.
+
+I really like the way everything comes together nicely when you have found
+the right way to combine values. The experience of refactoring haskell code,
+especially with hlint by my side hasn't been matched in any other language
+for now.
+
+[^1]: Don't get me started on why everyone uses `>>=` instead of the clearly
```
Comment (i-am-tom) [inline posts/2018-12-27-a-tale-of-servant-clients.md]: So much agreement here; does no one appreciate symmetry?!

File: posts/2018-12-27-a-tale-of-servant-clients.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+runClient f = do
+  ApplicationEnv{..} <- ask
+  liftIO $ runClientM (f $ newClient auth) clientEnv
+
+-- | This helper is there purely to improve readability.
+-- We could also define other versions with different arities
+-- (even though chaining multiple calls to @WithParam@ would
+-- work as well
+withParam :: a -> (a -> b) -> b
+withParam = flip ($)
```
Comment (i-am-tom) [inline posts/2018-12-27-a-tale-of-servant-clients.md]: I like this! I'm a big fan of `Data.Function.&` for similar reasons, particularly relating to your example below with `>>>`.

File: posts/2018-12-27-a-tale-of-servant-clients.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+```
+
+With all this done, and thanks to `(>>>)` from `Control.Category`,
+we can improve readability:
+
+```haskell
+editUserIO :: UserId -> UserData -> Application User
+editUserIO userId userData = runWithAuth $
+    withUser >>> withParam userId >>> editUser >>> withParam userData
```
Comment (i-am-tom) [inline posts/2018-12-27-a-tale-of-servant-clients.md]: So beautiful!

File: posts/2018-12-27-a-tale-of-servant-clients.md (hunk 1 above)
Comment (divarvel) [reply posts/2018-12-27-a-tale-of-servant-clients.md]: especially when mixed with `<$>` and `$`, the reading order switches from R-t-L to L-t-R, and without knowing precedence rules well, it's hard to parse it correctly just by reading it.

File: posts/2018-12-27-a-tale-of-servant-clients.md (hunk 2 above)
Comment (divarvel) [reply posts/2018-12-27-a-tale-of-servant-clients.md]: Sadly I couldn't mix `&` with `>>>` :-(

File: posts/2018-12-27-a-tale-of-servant-clients.md (hunk 3 above)
Comment (divarvel) [reply posts/2018-12-27-a-tale-of-servant-clients.md]: :-)
